File: Parser.hs
module Parser (AST(..), parser) where

import Lexer
import Control.Monad.State

-- Uppercase: AST nodes
data AST = Var String 
         | Let String AST AST
         | App AST    AST
         | Abs String AST
         | Empty

instance Show AST where
    show ast = showTree 0 ast where
        padding i = concat (replicate i "   ")
        showTree i (Var token)           = padding i ++ show (VAR token)
        showTree i (Let token ast1 ast2) = padding i ++ "LET " ++ show (VAR token) ++ "\n" ++ showTree (i+1) ast1 ++ "\n" ++ showTree (i+1) ast2
        showTree i (App ast1  ast2)      = padding i ++ "APP\n" ++ showTree (i+1) ast1 ++ "\n" ++ showTree (i+1) ast2
        showTree i (Abs token ast)       = padding i ++ "ABS " ++ show (VAR token) ++ "\n" ++ showTree (i+1) ast

parser :: [Token] -> AST
parser tokens = evalState (parseExpr (head tokens)) tokens
-- parser list = Let "abc" (App (Abs "def" (Var "ghi")) (Var "jkl"))

nextToken :: State [Token] Token
nextToken = do (t:ts) <- get
               if t == EOF then return EOF
               else put ts >> return (head ts)

currentToken :: State [Token] Token
currentToken = do t <- get
                  return (head t)

--------------- EXPR -----------
parseExpr :: Token -> State [Token] AST
parseExpr t = case t of
    (VAR str) -> parseApp t
    LPAR      -> parseApp t
    RPAR      -> error "Got ) at start of <EXPR>, should be / \\ ( or Var"
    SLASH     -> parseLet t
    BSLASH    -> parseAbs t
    EOF       -> error "Got EOF in <EXPR>"

---------------- LET ---------------
parseLet :: Token -> State [Token] AST
parseLet SLASH = do var  <- (nextToken >>= parseVar)
                    val  <- (currentToken >>= parseVal)
                    t <- currentToken
                    if t == EOF then error "Missing Expression after <LET>-Statemet."
                    else do expr <- parseExpr t
                            return (Let var val expr)
parseLet t = error ("Expected / (for <LET>), but got " ++ show t)

----------------- VAR -----------------
parseVar :: Token -> State [Token] String
parseVar (VAR str) = nextToken >> return str
parseVar t         = error ("Expector Var (for <VAR>), but got " ++ show t)

------------------- VAL ---------------------
parseVal :: Token -> State [Token] AST
parseVal (VAR str) = parseVar (VAR str) >>= (\x -> return (Var x))
parseVal LPAR      = do ast <- (nextToken >>= parseExpr)
                        t <- currentToken
                        if t /= RPAR then error ("Exepected ) (for <VAL>), but got " ++ show t)
                          else (nextToken >> return ast)
parseVal t         = error ("Expected Var or ( (for <VAL>), but got " ++ show t)

-------------------- APP ------------------------
isAppStart :: Token -> Bool
isAppStart (VAR str) = True
isAppStart LPAR      = True
isAppStart t         = False

parseAppAST :: AST -> Token -> State [Token] AST
parseAppAST ast t = if isAppStart t
                    then do val <- parseVal t
                            currentToken >>= parseAppAST (App ast val)
                    else return ast

parseApp :: Token -> State [Token] AST
parseApp t = do val <- parseVal t
                t <- currentToken
                if isAppStart t 
                  then do val2 <- parseVal t
                          currentToken >>= parseAppAST (App val val2)
                  else return val

----------------------- ABS ---------------------------
parseAbs :: Token -> State [Token] AST
parseAbs BSLASH = do var <- (nextToken >>= parseVar)
                     expr <- (currentToken >>= parseExpr)
                     return (Abs var expr)
parseAbs t = error ("Expected \\ (for <ABS>), but got " ++ show t)

File: cpp-interpreter/src/parser.cpp
#include "parser.h"
#include "sbsexception.h"

#include <iostream>

/*
Grammar:

EXPR = APP | LET | ABS
LET  = / VAR VAL EXPR
VAR  = v
VAL  = ( EXPR ) | VAR
APP  = VAL VAL APP' | VAL
APP' = VAL APP' | #
ABS  = \ VAR EXPR

first(EXPR) = ( v / \
first(APP) = ( v
first(LET) = /
first(ABS) = \
first(VAR) = v
first(VAL) = ( v

first(APP') = ( v follow(APP)


follow(EXPR) = ) #
follow(LET)  = ) #
follow(VAR)  = ) # ( v / \
follow(VAL)  = ) # ( v / \
follow(APP)  = ) #
follow(APP') = ) #
follow(ABS)  = ) #

*/

void parseError(const Token& token, const std::string& symbol, const std::string& expected) {
    throw SBSException(SBSException::Origin::PARSER, "Got Token " + token.toString() + " in " + symbol + ", but expected one of " + expected + ".", token.file, token.line, token.pos);
}


std::shared_ptr<AST> parseVal(const std::vector<Token>& tokens, int& i);
std::shared_ptr<AST> parseExpr(const std::vector<Token>& tokens, int& i);
std::shared_ptr<AST> parseLet(const std::vector<Token>& tokens, int& i);
std::shared_ptr<AST> parseAbs(const std::vector<Token>& tokens, int& i);
std::shared_ptr<AST> parseApp(const std::vector<Token>& tokens, int& i);


std::shared_ptr<AST> parse(const std::vector<Token>& tokens) {
    int i = 0;
    return parseExpr(tokens, i);
}


// VAL  = ( EXPR ) | VAR
std::shared_ptr<AST> parseVal(const std::vector<Token>& tokens, int& i) {
    if (tokens[i].type == Token::Type::VAR) {
        auto ast = std::make_shared<AST>(AST::Type::VAR, tokens[i].file, tokens[i].line, tokens[i].pos, tokens[i].fromLib);
        ast->var.name = tokens[i].value;

        i++;
        return ast;
    }
    else if (tokens[i].type == Token::Type::LPAR) {
        i++;

        auto ast = parseExpr(tokens, i);

        if(tokens[i].type != Token::Type::RPAR)
            parseError(tokens[i], "VAR", "\\");
        i++;

        return ast;
    }
    else parseError(tokens[i], "VAL", "[val] (");
    return nullptr;
}

// EXPR = APP | LET | ABS
std::shared_ptr<AST> parseExpr(const std::vector<Token>& tokens, int& i) {
    switch (tokens[i].type) {
        case Token::Type::LPAR:
        case Token::Type::VAR:
            return parseApp(tokens, i);
        case Token::Type::SLASH:
            return parseLet(tokens, i);
        case Token::Type::BSLASH:
            return parseAbs(tokens, i);
        case Token::Type::RPAR:
        case Token::Type::END:
            throw SBSException(SBSException::Origin::PARSER, "Got " + tokens[i].toString() + " in EXPR context.", tokens[i].file, tokens[i].line, tokens[i].pos);
    }
    parseError(tokens[i], "EXPR", "[val] ( / \\");
    return nullptr;
}

// LET  = / VAR VAL EXPR
std::shared_ptr<AST> parseLet(const std::vector<Token>& tokens, int& i) {
    auto ast = std::make_shared<AST>(AST::Type::LET, tokens[i].file, tokens[i].line, tokens[i].pos, tokens[i].fromLib);

    if(tokens[i].type != Token::Type::SLASH)
        parseError(tokens[i], "LET", "/");
    i++;

    if(tokens[i].type != Token::Type::VAR)
        parseError(tokens[i], "LET", "[val]");
    ast->let.name = tokens[i].value;
    i++;

    ast->let.value = parseVal(tokens, i);
    ast->let.next = parseExpr(tokens, i);

    return ast;
}

// ABS  = \ VAR EXPR
std::shared_ptr<AST> parseAbs(const std::vector<Token>& tokens, int& i) {
    auto ast = std::make_shared<AST>(AST::Type::ABS, tokens[i].file, tokens[i].line, tokens[i].pos, tokens[i].fromLib);

    if(tokens[i].type != Token::Type::BSLASH)
        parseError(tokens[i], "ABS", "\\");
    i++;

    if(tokens[i].type != Token::Type::VAR)
        parseError(tokens[i], "ABS", "[val]");
    ast->abs.name = tokens[i].value;
    i++;

    ast->abs.ast = parseExpr(tokens, i);

    return ast;
}

// APP  = VAL VAL APP' | VAL
// APP' = VAL APP' | #
std::shared_ptr<AST> parseApp2(std::shared_ptr<AST> last, const std::vector<Token>& tokens, int& i) {
    if((tokens[i].type == Token::Type::VAR) || (tokens[i].type == Token::Type::LPAR)) {
        auto ast = std::make_shared<AST>(AST::Type::APP, tokens[i].file, tokens[i].line, tokens[i].pos, tokens[i].fromLib);
        ast->app.first = std::move(last);
        ast->app.second = parseVal(tokens, i);
        return parseApp2(std::move(ast), tokens, i);
    }
    else return last;
}

std::shared_ptr<AST> parseApp(const std::vector<Token>& tokens, int& i) {
    auto first = parseVal(tokens, i);

    return parseApp2(std::move(first), tokens, i);
}

std::string padding(int i) {
    std::string s;
    for(int j = 0; j < i; j++)
        s += "  ";
    return s;
}

std::string ASTToString(const AST& ast, bool showLib, int i) {
    bool show = showLib || !ast.fromLib;

    if (ast.type == AST::Type::VAR)
        return padding(i) + ast.var.name;
    else if (ast.type == AST::Type::LET)
        return (show ? (padding(i) + "LET " + ast.let.name + "\n" + ASTToString(*ast.let.value, showLib, i+1) + "\n") : "") + ASTToString(*ast.let.next, showLib, i);
    else if (ast.type == AST::Type::APP)
        return padding(i) + "APP\n" + ASTToString(*ast.app.first, showLib, i+1) + "\n" + ASTToString(*ast.app.second, showLib, i+1);
    else if (ast.type == AST::Type::ABS)
        return padding(i) + "ABS " + ast.abs.name + "\n" + ASTToString(*ast.abs.ast, showLib, i+1);
    else return "NULLPTR";
}